NEXT-8593 - Customer impersonation

sobyte opened the pull request:
<!--
Thank you for contributing to Shopware! Please fill out this description template to help us to process your pull request.

Please make sure to fulfil our contribution guideline (https://developer.shopware.com/docs/resources/guidelines/code/contribution?category=shopware-platform-dev-en/contribution).

Do your changes need to be mentioned in the documentation?
Please create a second pull request at https://github.com/shopware/docs
-->

### 1. Why is this change necessary?
This PR continues the work of #3278.

### 2. What does this change do, exactly?
See #3278 

### 3. Describe each step to reproduce the issue or behaviour.
See #3278 

### 4. Please link to the relevant issues (if any).
* [NEXT-8593](https://issues.shopware.com/issues/NEXT-8593)
* #878 

### 5. Checklist

- [x] I have rebased my changes to remove merge conflicts
- [ ] I have written tests and verified that they fail without my change
- [x] I have created a [changelog file](https://github.com/shopware/platform/blob/trunk/adr/2020-08-03-implement-new-changelog.md) with all necessary information about my changes
- [ ] I have written or adjusted the documentation according to my changes
- [ ] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.


[NEXT-8593]: https://shopware.atlassian.net/browse/NEXT-8593?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

ugurkankya committed:
wip add first commit for testing

ugurkankya committed:
remove test console.log

ugurkankya committed:
[NEXT-8593] add login as customer from administration

ugurkankya committed:
[NEXT-8593] introduce InvalidCustomerIdException

ugurkankya committed:
[NEXT-8593] remove unused method from store context api service

ugurkankya committed:
[NEXT-8593] add error message if login as customer failed

ugurkankya committed:
[NEXT-8593] add missing changelog file

ugurkankya committed:
[NEXT-8593] fix phpunit tests

ugurkankya committed:
[NEXT-8593] feat: add modal with ability to select the sales channel and perform the login

ugurkankya committed:
[NEXT-8593] add login as customer route with moved logic

ugurkankya committed:
[NEXT-8593] change phpdoc comment of AbstractLoginAsCustomerRoute explain what it is exactly doing

ugurkankya committed:
[NEXT-8593] add unit test for LoginAsCustomerTokenGenerator

ugurkankya committed:
[NEXT-8593] update unreleased changelog add new files

ugurkankya committed:
[NEXT-8593] add new method customerNotFoundById to CustomerException

ugurkankya committed:
[NEXT-8593] add new constants for salesChannelId and customerId and token

ugurkankya committed:
[NEXT-8593] add pre set salesChannelId if id from sales channel context was found

ugurkankya committed:
Merge branch 'trunk' of github.com:ugurkankya/platform into next-8593/allow-admin-customer-login

ugurkankya committed:
fix login as customer modal not being loaded

ugurkankya committed:
fix wrong return type declaration in abstract route

ugurkankya committed:
wip restore accidentally deleted branch and add todo

ugurkankya committed:
Merge branch 'trunk' of github.com:ugurkankya/platform into next-8593/allow-admin-customer-login

Commit:
Merge branch 'trunk' into next-8593/allow-admin-customer-login

Commit:
[NEXT-8593] Apply suggestions. Fix errors from auto-merge. Optimize code.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/shopware/shopware?pullRequest=3713) <br/>All committers have signed the CLA.

Commit:
Merge branch 'trunk' into next-8593/allow-admin-customer-login

Commit:
[NEXT-8593] Fix eslint errors.

Commit:
[NEXT-8593] Fix phpunit tests.

Commit:
[NEXT-8593] Fix changelog.

Commit:
[NEXT-8593] Adjust wording for snippets.

ugurkankya commented:
@sobyte Looks good to me thanks

ugurkankya commented:
@shyim Can you have a look at this? Would be happy to see this merged :100:

AydinHassan commented:
@sobyte thanks for the PR. Is it ready for review? We will most likely do that next week as much of the team is away.

sobyte commented:
> @sobyte thanks for the PR. Is it ready for review? We will most likely do that next week as much of the team is away.

Yes, I think so. I've implemented the requested changes from @shyim of #3278 and added additional changes. I will try to fix the remaining errors in the pipeline as soon as possible, but that shouldn't block your review, does it?

AydinHassan commented:
@sobyte yep that's fine, just didn't want to prematurely review a large piece of work if it was still a wip :)

akf-bw commented on src/Core/Checkout/Customer/SalesChannel/LoginAsCustomerRoute.php:
```diff
@@ -0,0 +1,100 @@
+
+        $newToken = $restoredCart->getToken();
+
+        $event = new CustomerLoginEvent($context, $customer, $newToken);
```
@sobyte @AydinHassan
Do you think we should/could also add a new "AsCustomerLoginEvent" here so that we can differentiate/track whether the login is triggered by a standard customer login or an admin customer login event?
This would help create custom logic like an “Admin-Customer Login Log” to log when a user accesses a customer account.

To do this, it would also be nice to somehow get the user for whom the token was generated here to pass the admin user to this new event so that we can identify which user has logged into this customer account

shyim replied:
or a flag inside the event?

aragon999 replied:
Maybe also the `AccountService::loginById` method should be used to login the customer: https://github.com/shopware/shopware/blob/trunk/src/Core/Checkout/Customer/SalesChannel/AccountService.php#L93

As otherwise it would be possible to login to a customer, which is for example bound to a different sales channel. But in general I think it would be useful to have different events here (not sure if it would be sufficient to only have the customer id in that event).

shyim commented on src/Core/Checkout/Customer/SalesChannel/LoginAsCustomerRoute.php:
```diff
@@ -0,0 +1,100 @@
+    #[Route(path: '/store-api/account/login/customer', name: 'store-api.account.login-as-customer', methods: ['POST'])]
+    public function loginAsCustomer(RequestDataBag $data, SalesChannelContext $context): ContextTokenResponse
+    {
+        // TODO: find better way to handle this
```
You can use here the DataValidator which uses under the hood Symfony Validator

https://github.com/shopware/shopware/blob/66e20695afc4e6fd7c48ac587e0acc81f9d04486/src/Core/Content/Newsletter/SalesChannel/NewsletterUnsubscribeRoute.php#L91-L92

shyim replied:
This would also unify the error message when something is missing

akf-bw commented on src/Core/Checkout/Customer/LoginAsCustomerTokenGenerator.php:
```diff
@@ -0,0 +1,45 @@
+        $this->appSecret = $appSecret;
+    }
+
+    public function generate(string $salesChannelId, string $customerId): string
```
```suggestion
    public function generate(string $salesChannelId, string $customerId, string $userId): string
```

akf-bw commented on src/Core/Checkout/Customer/LoginAsCustomerTokenGenerator.php:
```diff
@@ -0,0 +1,45 @@
+        $data = json_encode($tokenData);
+
+        if ($data === false) {
+            throw InvalidLoginAsCustomerTokenException::invalidToken($salesChannelId . ':' . $customerId);
```
```suggestion
            throw InvalidLoginAsCustomerTokenException::invalidToken($salesChannelId . ':' . $customerId . ':' . $userId);
```

akf-bw commented on src/Core/Checkout/Customer/LoginAsCustomerTokenGenerator.php:
```diff
@@ -0,0 +1,45 @@
+        return hash_hmac('sha256', $data, $this->appSecret);
+    }
+
+    public function validate(string $givenToken, string $salesChannelId, string $customerId): void
```
```suggestion
    public function validate(string $givenToken, string $salesChannelId, string $customerId, string $userId): void
```

akf-bw commented on src/Core/Checkout/Customer/LoginAsCustomerTokenGenerator.php:
```diff
@@ -0,0 +1,45 @@
+
+    public function validate(string $givenToken, string $salesChannelId, string $customerId): void
+    {
+        $expectedToken = $this->generate($salesChannelId, $customerId);
```
```suggestion
        $expectedToken = $this->generate($salesChannelId, $customerId, $userId);
```

akf-bw commented on src/Core/Checkout/Customer/LoginAsCustomerTokenGenerator.php:
```diff
@@ -0,0 +1,45 @@
+    {
+        $tokenData = [
+            'salesChannelId' => $salesChannelId,
+            'customerId' => $customerId,
```
```suggestion
            'customerId' => $customerId,
            'userId' => $userId,
```

akf-bw commented on src/Core/Checkout/Customer/SalesChannel/LoginAsCustomerRoute.php:
```diff
@@ -0,0 +1,100 @@
+            throw CustomerException::missingToken();
+        }
+
+        $customerId = (string) $data->get(self::CUSTOMER_ID);
```
```suggestion
        $customerId = (string) $data->get(self::CUSTOMER_ID);
        $userId = (string) $data->get(self::USER_ID);
```

akf-bw commented on src/Core/Checkout/Customer/SalesChannel/LoginAsCustomerRoute.php:
```diff
@@ -0,0 +1,100 @@
+#[Package('checkout')]
+class LoginAsCustomerRoute extends AbstractLoginAsCustomerRoute
+{
+    public const CUSTOMER_ID = 'customerId';
```
```suggestion
    public const CUSTOMER_ID = 'customerId';
    
    public const USER_ID = 'userId';
```

akf-bw commented on src/Core/Framework/Api/Controller/SalesChannelProxyController.php:
```diff
@@ -135,6 +138,47 @@ public function assignCustomer(Request $request, Context $context): Response
+            throw ApiException::invalidSalesChannelId($salesChannelId);
+        }
+
+        $token = $this->loginAsCustomerTokenGenerator->generate($salesChannelId, $customerId);
```
```suggestion
        $source = $context->getSource();
        if (!$source instanceof AdminApiSource) {
            throw new InvalidContextSourceException(AdminApiSource::class, $context->getSource()::class);
        }
        $userId = $source->getUserId();
        
        $token = $this->loginAsCustomerTokenGenerator->generate($salesChannelId, $customerId, $userId);
```

akf-bw commented on src/Storefront/Controller/AuthController.php:
```diff
@@ -312,4 +317,42 @@ public function resetPassword(RequestDataBag $data, SalesChannelContext $context
         return $this->redirectToRoute('frontend.account.profile.page');
     }
+
+    #[Route(path: '/account/login/customer/{token}/{salesChannelId}/{customerId}', name: 'frontend.account.login.customer', methods: ['GET'])]
```
```suggestion
    #[Route(path: '/account/login/customer/{token}/{salesChannelId}/{customerId}/{userId}', name: 'frontend.account.login.customer', methods: ['GET'])]
```

akf-bw commented on src/Storefront/Controller/AuthController.php:
```diff
@@ -312,4 +317,42 @@ public function resetPassword(RequestDataBag $data, SalesChannelContext $context
+            $dataBag = new RequestDataBag([
+                LoginAsCustomerRoute::TOKEN => $token,
+                LoginAsCustomerRoute::SALES_CHANNEL_ID => $salesChannelId,
+                LoginAsCustomerRoute::CUSTOMER_ID => $customerId,
```
```suggestion
                LoginAsCustomerRoute::CUSTOMER_ID => $customerId,
                LoginAsCustomerRoute::USER_ID => $userId,
```

akf-bw commented on src/Storefront/Controller/AuthController.php:
```diff
@@ -312,4 +317,42 @@ public function resetPassword(RequestDataBag $data, SalesChannelContext $context
     }
+
+    #[Route(path: '/account/login/customer/{token}/{salesChannelId}/{customerId}', name: 'frontend.account.login.customer', methods: ['GET'])]
+    public function loginAsCustomer(string $token, string $salesChannelId, string $customerId, SalesChannelContext $context, Request $request): Response
```
```suggestion
    public function loginAsCustomer(string $token, string $salesChannelId, string $customerId, string $userId, SalesChannelContext $context, Request $request): Response
```

akf-bw commented on src/Core/Framework/Api/Controller/SalesChannelProxyController.php:
```diff
@@ -135,6 +138,47 @@ public function assignCustomer(Request $request, Context $context): Response
+            $this->generateUrl('frontend.account.login.customer', [
+                'token' => $token,
+                'salesChannelId' => $salesChannelId,
+                'customerId' => $customerId,
```
```suggestion
                'customerId' => $customerId,
                'userId' => $userId,
```

akf-bw commented on src/Core/Checkout/Customer/SalesChannel/LoginAsCustomerRoute.php:
```diff
@@ -0,0 +1,100 @@
+        $customerId = (string) $data->get(self::CUSTOMER_ID);
+        $token = (string) $data->get(self::TOKEN);
+
+        $this->tokenGenerator->validate($token, $context->getSalesChannelId(), $customerId);
```
```suggestion
        $this->tokenGenerator->validate($token, $context->getSalesChannelId(), $customerId, $userId);
```

akf-bw commented on src/Core/Checkout/Customer/SalesChannel/LoginAsCustomerRoute.php:
```diff
@@ -0,0 +1,100 @@
+
+        $customer = $this->fetchCustomer($customerId, $context->getContext());
+
+        $event = new CustomerBeforeLoginEvent($context, $customer->getEmail());
```
```suggestion
        $event = new CustomerBeforeLoginEvent($context, $customer->getEmail(), $userId);
```

akf-bw commented on src/Core/Checkout/Customer/SalesChannel/LoginAsCustomerRoute.php:
```diff
@@ -0,0 +1,100 @@
+
+        $newToken = $restoredCart->getToken();
+
+        $event = new CustomerLoginEvent($context, $customer, $newToken);
```
```suggestion
        $event = new CustomerLoginEvent($context, $customer, $newToken, $userId);
```

akf-bw reviewed (changes requested):
We can pass the userId from the initial request to the final CustomerLoginEvent.
In the CustomerLoginEvent & CustomerBeforeLoginEvent we can add a new:
`?string $userId = null`

ghost commented:
@sobyte to continue as fast as possible i extended your PR into #3727.
I added all the additional comments from @shyim & @aragon999 to the new PR & reworked some parts of it.
Also all tests are now fixed.

ugurkankya commented:
> We can pass the userId from the initial request to the final CustomerLoginEvent. In the CustomerLoginEvent & CustomerBeforeLoginEvent we can add a new: `?string $userId = null`

Adding `$userId` makes sense to me but i think it should not be nullable i think, since the parameter will always be passed.

ghost commented:
> Adding `$userId` makes sense to me but i think it should not be nullable i think, since the parameter will always be passed.

I already implemented this in https://github.com/shopware/shopware/pull/3727. The userId must be nullable because the events and functions are also used for standard customer login where no admin user is involved.

Bird87ZA commented:
Superseded by https://github.com/shopware/shopware/pull/3727 as mentioned [here](https://github.com/shopware/shopware/pull/3713#issuecomment-2126889072).
